content/sharing-variables-with-stages-and-jobs.md

```diff
@@ -0,0 +1,222 @@
+Title: Gotchas when share variables with Azure DevOps stages and jobs
+Date: 2021-07-08
+Category: Posts
+Tags: azure-devops, pipelines, tips
+Slug: sharing-variables-with-stages-and-jobs
+Author: Willy-Peter Schaub
+Summary: Understanding and avoiding gotchas that may keep you up at 2AM, troubleshooting YAML-based Azure Pipelines
+
+I assume that you have read our [How to share variables amongst Azure Pipeline agents](/sharing-variables-with-stages-and-jobs.html), which shared a few turbulent moments we experienced while troubleshooting this feature earlier this year. With this post we continue our troubleshooting excursion to highlight a few gotchas that have caused lots of head scratching.
+
+---
+
+# Core Syntax
+
+> ![Syntax](/images/sharing-variables-with-stages-and-jobs-3.png)
+
+Here are two reminders to tattoo on your forearm, when referencing variables:
+
+- Within a stage, map variables as: ```dependencies.<stage>.<job>.outputs['<step>.<name>']```
+- Across stages, map variables as: ```stageDependencies.<stage>.<job>.outputs['<step>.<name>']```
+
+Let us lift the pipeline bonnet and explore.
+
+---
+
+# STEP 1: Define a variable to be shared
+
+We define three variables, named secretValue1, secretValue2, and secretValue3 in our job called StageOneJobOne. Note that we explicitly name two of the steps and leave one as default. Sounds simple, but this will bite is later on.
+
+```
+  - job: StageOneJobOne
+    pool:
+      vmImage: 'windows-latest'
+    steps:
+    - bash: echo "##vso[task.setvariable variable=secretValue1;isOutput=true]BINGO-1!"
+      name: SetVariable1
+    - bash: echo "##vso[task.setvariable variable=secretValue2;isOutput=true]BINGO-2!"
+    - bash: echo "##vso[task.setvariable variable=secretValue3;isOutput=true]BINGO-3!"
+      name: SetVariable3
+    - task: DutchWorkzToolsAllVariables@1
+```
+
+
+The last task, which unfortunately only runs on a Windows-based agent, uses the [Display all variables](https://marketplace.visualstudio.com/items?itemName=dutchworkz.DisplayAllVariables) extension, which displays all variables that are available during your Build or Release. It is an exceptional troubleshooting companion and we find all three of our variables listed as individual variables, and as part of the VSTS_PUBLIC_VARIABLES collection. 
```

Comment:
Hi @wpschaub, the "Display all variables" extension has been unpublished from Marketplace. I think we should avoid using it.
I use this simple script to log all the variables: `- bash: 'env | sort'`. It works on all agents.
Thanks